File: handler.py
#! /usr/bin/env python3

import pathlib
import json
import os
from traceback import format_exception
import typing as t

import kong_pdk.pdk.kong as kong

from pangea import PangeaConfig
from pangea.services.ai_guard import AIGuard
from pangea_translator import get_translator

token_secret = os.getenv("PANGEA_AI_GUARD_TOKEN", "")
with open(token_secret) as f: token = f.read()

Schema = (
    {"message": {"type": "string"}},
)
version = '0.1.0'
priority = 0

DEFAULT_PANGEA_DOMAIN = "aws.us.pangea.cloud"


class Operation:
    def __init__(self, op_params: dict):
        self.json = op_params.copy()
        if "recipe" not in self.json:
            self.json["recipe"] = "pangea_prompt_guard"


class Rule:
    def __init__(self, rule: dict):
        self.rule = rule
        self.host = rule.get("host")
        self.endpoint = rule.get("endpoint")
        self.prefix = rule.get("prefix")
        self.protos = rule.get("protocols")
        self.ports = rule.get("ports")
        self.parser = rule.get("parser")
        self.allow_failure = rule.get("allow_on_error", False)

    def match(self, host, endpoint, port, protocol, prefix) -> bool:
        if host != self.host:
            return False
        if self.endpoint and endpoint != self.endpoint:
            return False
        if self.prefix and prefix != self.prefix:
            return False
        if self.protos and protocol not in self.protos:
            return False
        if self.ports and port not in self.ports:
            return False
        return True

    def operation_params(self, op, service="ai_guard") -> t.Optional[Operation]:
        svc = self.rule.get(service)
        if not svc:
            return None
        info = svc.get(op, {}).get("parameters")
        if info is None or info.get("enabled", True) is not True:
            return None
        return Operation(info)


class PangeaKongConfig:
    def __init__(self, j: dict):
        self.domain = j.get("pangea_domain")
        self.insecure = j.get("insecure", False)
        self.header_recipe_map = j.get("headers", {})
        if not self.domain:
            self.domain = DEFAULT_PANGEA_DOMAIN
        rules = j.get("rules", [])
        self.rules: t.List[Rule] = []
        for i, rule in enumerate(rules):
            if not rule.get("host"):
                kong.kong.log.warn(f"Rule {i} is missing the host which is required. Ignoring rule")
                continue
            endpoint = rule.get("endpoint")
            prefix = rule.get("prefix")
            if not endpoint or not prefix:
                kong.kong.log.warn(f"Rule {i} is missing the endpoint or forwarding prefix, at least one is required. Ignoring rule")
                continue
            self.rules.append(Rule(rule))

    def match_rule(self, k: kong.kong) -> t.Optional[Rule]:
        proto, err = k.request.get_forwarded_scheme()
        if err:
            k.log.err(f"failed to get scheme: {err}")
            return None
        host, err = k.request.get_forwarded_host()
        if err:
            k.log.err(f"failed to get host: {err}")
            return None
        port, err = k.request.get_forwarded_port()
        if err:
            k.log.err(f"failed to get port: {err}")
            return None
        endpoint, err = k.request.get_forwarded_path()
        if err:
            k.log.err(f"failed to get endpoint: {err}")
            return None
        prefix, err = k.request.get_forwarded_prefix()
        if err:
            prefix = None
        if prefix and endpoint.startswith(prefix):
            endpoint = endpoint[len(prefix):]
        for rule in self.rules:
            if rule.match(host, endpoint, port, proto, prefix):
                return rule

        return None

def load_config():
    loc = os.getenv("PANGEA_KONG_CONFIG_FILE")
    if loc:
        pth = pathlib.Path(loc)
    else:
        pth = pathlib.Path("/etc/pangea_kong_config.json")
    if pth.exists():
        json_config = json.load(open(pth))
        return PangeaKongConfig(json_config)
    else:
        kong.kong.log.warn(f"No config provided, using default")
        return PangeaKongConfig(default_config)


default_config = {
  "pangea_domain": DEFAULT_PANGEA_DOMAIN,
  "rules": [
    {
      "host": "api.openai.com",
      "endpoint": "/v1/chat/completions",
      # "prefix": "/a_prefix",
      "allow_on_error": False,
      "protocols": ["https"],
      "ports": ["443"],
      "audit_values": {
        "model": "openai"
      },
      "ai_guard": {
        "request": {
          "parameters": {
            "recipe": "pangea_prompt_guard",
          }
        },
        "response": {
        }
      }
    }
  ]
}


config = load_config()


class Plugin:
    def __init__(self, k_config):
        self.k_config = k_config
        kwargs = {
            "domain": config.domain,
        }
        if not config.domain.endswith(".pangea.cloud"):
            kwargs["environment"] = "local"
        if config.insecure:
            kwargs["insecure"] = True

        self.ai_guard = AIGuard(token, config=PangeaConfig(**kwargs))

    def access(self, k: kong.kong):
        allow_failure = False
        try:
            rule = config.match_rule(k)
            if not rule:
                k.log.debug(f"No rule matched {k.request.get_host()}{k.request.get_path()}, allowing")
                return
            allow_failure = rule.allow_failure is True
            op = rule.operation_params("request")
            k.log.debug(f"Rule op: {json.dumps(op.json)}")
            if op is None:
                k.log.debug(f"No work for 'request', allowing")
                return

            k.log.debug(f"Request headers: {k.request.get_headers()}")
            # recipe is in the config, but can be overridden by this header
            recipe = k.request.get_header("x-pangea-aig-recipe")
            if recipe and recipe[0]:
                # it comes in an array
                recipe = recipe[0]
            else:
                recipe = None
            # can be further overridden by the configured header/recipe map
            # for header_name, recipe_map in config.header_recipe_map.items():
            #     header_name = header_name.lower()
            #     header = k.request.get_header(header_name)
            #     if not header:
            #         continue
            #     else:
            #         header = header[0]
            #     if header in recipe_map:
            #         recipe = recipe_map[header]
            if recipe:
                op.json["recipe"] = recipe

            text, err = k.request.get_raw_body()
            if err:
                k.log.err(f"Got error when getting request raw body: '{err}'")
            translator = None
            log_fields = {
                "model": f'{{"provider": "{rule.parser}"}}',
                "extra_info": f'{{"api": "{k.request.get_forwarded_path()[0]}"}}',
            }
            op.json["log_fields"] = log_fields
            try:
                if isinstance(text, bytes):
                    text = text.decode("utf8")
                payload = json.loads(text)
                k.log.debug(f"PARAMS: {json.dumps(op.json)}")
                if rule.parser:
                    translator = get_translator(payload, llm_hint=rule.parser)
                    model, model_version = translator.get_model_and_version()
                    if not model_version:
                        model_version = "null"
                    log_fields["model"] = f'{{"provider": "{rule.parser}", "model": "{model}", "version": {model_version}}}'
                    pangea_messages = translator.get_pangea_messages()
                    response = self.ai_guard.guard_text(messages=pangea_messages.messages, **op.json)
                else:
                    response = self.ai_guard.guard_text(messages=payload, **op.json)
            except (json.JSONDecodeError, TypeError) as e:
                k.log.debug(f"JSON parse failed: {str(e)}")
                k.log.debug(f"JSON parse failed: {repr(text)}")
                response = self.ai_guard.guard_text(str(text), **op.json)

            if response.http_status != 200:
                k.log.err(f"Failed to call AI Guard: {response.status_code}, {response.text}")
                raise Exception("Failed to call AI Guard")

            new_prompt = response.json["result"].get("prompt_text")
            if not new_prompt:
                new_prompt = response.json["result"].get("prompt_messages")
                if rule.parser and translator:
                    new_prompt = translator.transformed_original_input(messages=new_prompt)

                new_prompt = json.dumps(new_prompt)
            else:
                new_prompt = str(new_prompt)

            blocked = response.json["result"].get("blocked", False)
            if blocked:
                for name, result in response.json["result"]["detectors"].items():
                    if result.get("detected", False):
                        k.log.warn(f"Detected unwanted prompt characteristics: {name}, {json.dumps(response.json)}")
                        return k.response.error(400, f"Prompt has been rejected: {response.json['summary']}", {"Content-Type": "text/html"})
            else:
                k.log.debug(f"Prompt allowed: {json.dumps(response.json)}")
                if new_prompt:
                    k.service.request.set_raw_body(new_prompt)
        except Exception as e:
            k.log.err(f"Exception while trying to call AI Guard:\n {format_exception(e)}")
            if not allow_failure:
                k.response.error(400, "Prompt has been rejected", {"Content-Type": "text/html"})


# for running in a dedicated process
if __name__ == "__main__":
    from kong_pdk.cli import start_dedicated_server
    start_dedicated_server("pangea_kong", Plugin, version, priority, Schema)

File: packages/lua/pangea-ai-guard/kong/plugins/pangea-ai-guard/handler.lua
local cjson = require "cjson"
local http = require "resty.http"
local translator = require "kong.plugins.pangea-ai-guard.pangea-translator.translate"

local DEFAULT_PANGEA_DOMAIN = "aws.us.pangea.cloud"


-- Rule class
local Rule = {}
Rule.__index = Rule

function Rule.new(rule)
    local self = setmetatable({}, Rule)
    self.rule = rule
    self.host = rule.host
    self.endpoint = rule.endpoint
    self.prefix = rule.prefix
    self.protos = rule.protocols
    self.ports = rule.ports
    self.parser = rule.parser
    self.allow_failure = rule.allow_on_error or false
    return self
end

function Rule:match(host, endpoint, port, protocol, prefix)
    if host ~= self.host then
        return false
    end
    if self.endpoint and endpoint ~= self.endpoint then
        return false
    end
    if self.prefix and prefix ~= self.prefix then
        return false
    end
    if self.protos and not self.protos[protocol] then
        return false
    end
    if self.ports and not self.ports[tostring(port)] then
        return false
    end
    return true
end

function Rule:operation_params(op, service)
    service = service or "ai_guard"
    local svc = self.rule[service]
    if not svc then
        kong.log.info("PangeaPlugin: Rule:operation_params service not found")
        return nil
    end
    local info = svc[op] and svc[op].parameters
    if not info or info.enabled == false then
        kong.log.info("PangeaPlugin: Rule:operation_params operation or parameters not found")
        return nil
    end
    return info
end

-- Configuration class
local PangeaKongConfig = {}
PangeaKongConfig.__index = PangeaKongConfig

function PangeaKongConfig.new(config)
    local self = setmetatable({}, PangeaKongConfig)
    self.domain = config.pangea_domain or DEFAULT_PANGEA_DOMAIN
    self.insecure = config.insecure or false
    self.header_recipe_map = config.headers or {}
    self.rules = {}

    for _, rule in ipairs(config.rules or {}) do
        if rule.host then
            local endpoint = rule.endpoint
            local prefix = rule.prefix
            if endpoint and prefix then
                table.insert(self.rules, Rule.new(rule))
            end
        end
    end

    return self
end

function PangeaKongConfig:match_rule()
    local proto = kong.request.get_forwarded_scheme()
    local host = kong.request.get_forwarded_host()
    local port = kong.request.get_forwarded_port()
    local endpoint = kong.request.get_forwarded_path()
    local prefix = kong.request.get_forwarded_prefix()

    if prefix and endpoint:sub(1, #prefix) == prefix then
        endpoint = endpoint:sub(#prefix + 1)
    end

    for _, rule in ipairs(self.rules) do
        if rule:match(host, endpoint, port, proto, prefix) then
            return rule
        end
    end

    return nil
end

-- Load configuration
local function load_config()
    kong.log.debug("PangeaPlugin: load_config")
    local config_file = os.getenv("PANGEA_KONG_CONFIG_FILE") or "/etc/pangea_kong_config.json"
    local file = io.open(config_file, "r")

    kong.log.debug("PangeaPlugin: load_config. ", "config_file: ", config_file)
    if file then
        local content = file:read("*all")
        file:close()
        local config = cjson.decode(content)
        kong.log.debug("PangeaPlugin: load_config. ", "config: ", cjson.encode(config))
        return PangeaKongConfig.new(config)
    end

    -- Default config
    local default_config = {
        pangea_domain = DEFAULT_PANGEA_DOMAIN,
        rules = {
            {
                host = "localhost",
                endpoint = "/v1/chat/completions",
                allow_on_error = false,
                protocols = {https = true},
                ports = {["443"] = true},
                audit_values = {
                    model = "openai"
                },
                ai_guard = {
                    request = {
                        parameters = {
                            recipe = "pangea_prompt_guard"
                        }
                    },
                    response = {}
                }
            }
        }
    }
    return PangeaKongConfig.new(default_config)
end

-- Plugin class
local PangeaAIGuardHandler = {
    PRIORITY = 1000,
    VERSION = "0.0.1",
}

-- PangeaAIGuardHandler.__index = PangeaAIGuardHandler

function PangeaAIGuardHandler:init()
    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler.init")
    -- local self = setmetatable({}, Plugin)
    self.token = os.getenv("PANGEA_AI_GUARD_TOKEN")
    if self.token == nil or self.token == "" then
        error("PangeaAIGuardHandler.init ERROR: PANGEA_AI_GUARD_TOKEN environment variables is required")
    end
    self.config = load_config()
end

function PangeaAIGuardHandler:init_worker()
  PangeaAIGuardHandler:init()
end

function PangeaAIGuardHandler:access()
    local allow_failure = false
    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler:access (start)")

    local rule = self.config:match_rule()
    if not rule then
        kong.log.info("PangeaPlugin: No rule matched ", kong.request.get_host(), kong.request.get_path(), ", allowing")
        return
    end

    allow_failure = rule.allow_failure
    local op = rule:operation_params("request")

    if not op then
        kong.log.info("PangeaPlugin: No work for 'request', allowing")
        return
    end

    -- Get recipe from header or config
    local recipe = kong.request.get_header("x-pangea-aig-recipe")
    if recipe and recipe[1] then
        op.recipe = recipe[1]
    end

    -- Get request body
    local body = kong.request.get_raw_body()
    if not body then
        kong.log.err("Failed to get request body")
        if not allow_failure then
            return kong.response.error(400, "Failed to get request body")
        end
        return
    end

    -- Prepare log fields
    local log_fields = {
        model = string.format('{"provider": "%s"}', rule.parser or ""),
        extra_info = string.format('{"api": "%s"}', kong.request.get_forwarded_path())
    }
    op.log_fields = log_fields

    -- Process request
    local success, payload = pcall(cjson.decode, body)
    if not success then
        kong.log.info("PangeaPlugin: JSON parse failed: ", payload)
        payload = body
    end

    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler:access. ", "payload: ", cjson.encode(payload))

    -- Handle translation if needed
    local translator_instance
    if rule.parser then
        translator_instance = translator.get_translator(payload, rule.parser)
      if translator_instance then
        local model, model_version = translator_instance:get_model_and_version()
        model_version = model_version or "null"
        log_fields.model = string.format('{"provider": "%s", "model": "%s", "version": %s}',
            rule.parser, model, model_version)
        payload = translator_instance:get_pangea_messages()
      else
        kong.log.err("PangeaPlugin: PangeaAIGuardHandler:access. ", "translator_instance not found for ", rule.parser)
      end
    end

    -- Make request to Pangea AI Guard
    local body = {
      messages = payload,
      recipe = op.recipe,
      log_fields = op.log_fields
    }
    local url = "https://ai-guard." .. self.config.domain .. "/v1/text/guard"
    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler:access", "Making request to ", url, ". body: ", cjson.encode(body))

    local httpc = http.new()
    local res, err = httpc:request_uri(url, {
        method = "POST",
        body = cjson.encode(body),
        headers = {
            ["Authorization"] = "Bearer " .. self.token,
            ["Content-Type"] = "application/json"
        }
    })

    if not res then
        kong.log.err("Failed to call AI Guard: ", err)
        if not allow_failure then
            return kong.response.error(400, "Failed to call AI Guard")
        end
        return
    end

    if res.status ~= 200 then
        kong.log.err("AI Guard returned error: ", res.status, " ", res.body)
        if not allow_failure then
            return kong.response.error(400, "AI Guard service error")
        end
        return
    end

    local response = cjson.decode(res.body)
    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler:access", "AI Guard response: ", cjson.encode(response))

    local new_prompt = response.result.prompt_text or response.result.prompt_messages
    kong.log.debug("PangeaPlugin: PangeaAIGuardHandler:access", "new_prompt: ", cjson.encode(new_prompt))

    if response.result.blocked then
        for name, result in pairs(response.result.detectors) do
            if result.detected then
                kong.log.warn("Detected unwanted prompt characteristics: ", name, " ", cjson.encode(response))
                return kong.response.error(400, "Prompt has been rejected: " .. response.summary)
            end
        end
    else
        kong.log.debug("PangeaPlugin: Prompt allowed")
        if new_prompt then
            if rule.parser and translator_instance then
                new_prompt = translator_instance:transformed_original_input(new_prompt)
            end
            kong.service.request.set_raw_body(cjson.encode(new_prompt))
        end
    end
end

return PangeaAIGuardHandler
